- This Exploratory Data Analysis (EDA) will focus on overall Data Analyst job title, job description, salary estimate

- Let's import the required libraries

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt 
import seaborn as sns 
import matplotlib as mpl


import plotly 
import plotly.express as px
import plotly.graph_objs as go
import plotly.offline as py
from plotly.offline import iplot
from plotly.subplots import make_subplots
import plotly.figure_factory as ff


import warnings
warnings.filterwarnings('ignore') 

### Overview Stage

- Read the csv
- Look for basic information about the dataset

In [2]:
df = pd.read_csv('../input/data-analyst-jobs/DataAnalyst.csv')

In [3]:
df.head()

Unnamed: 0                                          Job Title  \
0           0  Data Analyst, Center on Immigration and Justic...   
1           1                               Quality Data Analyst   
2           2  Senior Data Analyst, Insights & Analytics Team...   
3           3                                       Data Analyst   
4           4                             Reporting Data Analyst   

              Salary Estimate  \
0  $37K-$66K (Glassdoor est.)   
1  $37K-$66K (Glassdoor est.)   
2  $37K-$66K (Glassdoor est.)   
3  $37K-$66K (Glassdoor est.)   
4  $37K-$66K (Glassdoor est.)   

                                     Job Description  Rating  \
0  Are you eager to roll up your sleeves and harn...     3.2   
1  Overview\n\nProvides analytical and technical ...     3.8   
2  We’re looking for a Senior Data Analyst who ha...     3.4   
3  Requisition NumberRR-0001939\nRemote:Yes\nWe c...     4.1   
4  ABOUT FANDUEL GROUP\n\nFanDuel Group is a worl...     3.9   

                              Company Name      Location  Headquarters  \
0           Vera Institute of Justice\n3.2  New York, NY  New York, NY   
1  Visiting Nurse Service of New York\n3.8  New York, NY  New York, NY   
2                         Squarespace\n3.4  New York, NY  New York, NY   
3                            Celerity\n4.1  New York, NY    McLean, VA   
4                             FanDuel\n3.9  New York, NY  New York, NY   

                     Size  Founded               Type of ownership  \
0    201 to 500 employees     1961          Nonprofit Organization   
1        10000+ employees     1893          Nonprofit Organization   
2  1001 to 5000 employees     2003               Company - Private   
3    201 to 500 employees     2002  Subsidiary or Business Segment   
4   501 to 1000 employees     2009               Company - Private   

                           Industry                            Sector  \
0                 Social Assistance                        Non-Profit   
1  Health Care Services & Hospitals                       Health Care   
2                          Internet            Information Technology   
3                       IT Services            Information Technology   
4               Sports & Recreation  Arts, Entertainment & Recreation   

                      Revenue Competitors Easy Apply  
0  $100 to $500 million (USD)          -1       True  
1      $2 to $5 billion (USD)          -1         -1  
2    Unknown / Non-Applicable     GoDaddy         -1  
3   $50 to $100 million (USD)          -1         -1  
4  $100 to $500 million (USD)  DraftKings       True

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2253 entries, 0 to 2252
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         2253 non-null   int64  
 1   Job Title          2253 non-null   object 
 2   Salary Estimate    2253 non-null   object 
 3   Job Description    2253 non-null   object 
 4   Rating             2253 non-null   float64
 5   Company Name       2252 non-null   object 
 6   Location           2253 non-null   object 
 7   Headquarters       2253 non-null   object 
 8   Size               2253 non-null   object 
 9   Founded            2253 non-null   int64  
 10  Type of ownership  2253 non-null   object 
 11  Industry           2253 non-null   object 
 12  Sector             2253 non-null   object 
 13  Revenue            2253 non-null   object 
 14  Competitors        2253 non-null   object 
 15  Easy Apply         2253 non-null   object 
dtypes: float64(1), int64(2),

- Even though we have information about the salary ('Salary Estimate'), it is not in the numeric version, same for 'Size' and 'Revenue'

- Based on our research interest we can make arrangments on these variables


### Preparing dataset for the EDA

- Look for the missing values
- Look for the area of interest (based on the research question we have)
- Make adjustments

- Let's look at the missing values

In [5]:
df.isnull().sum()

Unnamed: 0           0
Job Title            0
Salary Estimate      0
Job Description      0
Rating               0
Company Name         1
Location             0
Headquarters         0
Size                 0
Founded              0
Type of ownership    0
Industry             0
Sector               0
Revenue              0
Competitors          0
Easy Apply           0
dtype: int64

- Do we have only 1 missing value ?

- Nope. Even we look at the top 5 rows of the data we can easily see (-1). 
- Most of the time in the real world datasets, unknown and missing values are not recorded like sofware recognizable version of the missing value. Usually, as a data person, we have to detect and deal with them.
- So let's make (-1) as a mising value

In [6]:
df.replace('-1', np.nan)

Unnamed: 0                                          Job Title  \
0              0  Data Analyst, Center on Immigration and Justic...   
1              1                               Quality Data Analyst   
2              2  Senior Data Analyst, Insights & Analytics Team...   
3              3                                       Data Analyst   
4              4                             Reporting Data Analyst   
...          ...                                                ...   
2248        2248  RQS - IHHA - 201900004460 -1q Data Security An...   
2249        2249              Senior Data Analyst (Corporate Audit)   
2250        2250  Technical Business Analyst (SQL, Data analytic...   
2251        2251                Data Analyst 3, Customer Experience   
2252        2252                        Senior Quality Data Analyst   

                  Salary Estimate  \
0      $37K-$66K (Glassdoor est.)   
1      $37K-$66K (Glassdoor est.)   
2      $37K-$66K (Glassdoor est.)   
3      $37K-$66K (Glassdoor est.)   
4      $37K-$66K (Glassdoor est.)   
...                           ...   
2248  $78K-$104K (Glassdoor est.)   
2249  $78K-$104K (Glassdoor est.)   
2250  $78K-$104K (Glassdoor est.)   
2251  $78K-$104K (Glassdoor est.)   
2252  $78K-$104K (Glassdoor est.)   

                                        Job Description  Rating  \
0     Are you eager to roll up your sleeves and harn...     3.2   
1     Overview\n\nProvides analytical and technical ...     3.8   
2     We’re looking for a Senior Data Analyst who ha...     3.4   
3     Requisition NumberRR-0001939\nRemote:Yes\nWe c...     4.1   
4     ABOUT FANDUEL GROUP\n\nFanDuel Group is a worl...     3.9   
...                                                 ...     ...   
2248  Maintains systems to protect data from unautho...     2.5   
2249  Position:\nSenior Data Analyst (Corporate Audi...     2.9   
2250  Title: Technical Business Analyst (SQL, Data a...    -1.0   
2251  Summary\n\nResponsible for working cross-funct...     3.1   
2252  You.\n\nYou bring your body, mind, heart and s...     3.4   

                                 Company Name        Location  \
0              Vera Institute of Justice\n3.2    New York, NY   
1     Visiting Nurse Service of New York\n3.8    New York, NY   
2                            Squarespace\n3.4    New York, NY   
3                               Celerity\n4.1    New York, NY   
4                                FanDuel\n3.9    New York, NY   
...                                       ...             ...   
2248                       Avacend, Inc.\n2.5      Denver, CO   
2249                   Arrow Electronics\n2.9  Centennial, CO   
2250                                 Spiceorb      Denver, CO   
2251         Contingent Network Services\n3.1  Centennial, CO   
2252                          SCL Health\n3.4  Broomfield, CO   

          Headquarters                    Size  Founded  \
0         New York, NY    201 to 500 employees     1961   
1         New York, NY        10000+ employees     1893   
2         New York, NY  1001 to 5000 employees     2003   
3           McLean, VA    201 to 500 employees     2002   
4         New York, NY   501 to 1000 employees     2009   
...                ...                     ...      ...   
2248    Alpharetta, GA     51 to 200 employees       -1   
2249    Centennial, CO        10000+ employees     1935   
2250               NaN                     NaN       -1   
2251  West Chester, OH    201 to 500 employees     1984   
2252    Broomfield, CO        10000+ employees     1864   

                   Type of ownership                                 Industry  \
0             Nonprofit Organization                        Social Assistance   
1             Nonprofit Organization         Health Care Services & Hospitals   
2                  Company - Private                                 Internet   
3     Subsidiary or Business Segment                              IT Services   
4       

- Are we done with the 'unrecognizable missing value detection job'
- Sorry, I don't think so.
- We have just dealt with str version of the (-1), but we can still see (-1) in the founded column, which is numeric type.
- So we have to deal with all possible version of the (-1)

In [7]:
df.replace(-1, np.nan, inplace=True)

In [8]:
df.replace('-1', np.nan, inplace=True)

In [9]:
df.replace(-1.0, np.nan, inplace=True)

- Now let's see whether we still have only one missing value

In [10]:
df.isnull().sum()

Unnamed: 0              0
Job Title               0
Salary Estimate         1
Job Description         0
Rating                272
Company Name            1
Location                0
Headquarters          172
Size                  163
Founded               660
Type of ownership     163
Industry              353
Sector                353
Revenue               163
Competitors          1732
Easy Apply           2173
dtype: int64

- Yes, that's the reality of the real life data. 

In this EDA, we will focus on:
   - Job Title
   - Salary Estimate
   - Job Description
   - Rating
   - Company name
   - Industry
   - Sector

- Let's look at the each variable in detail and make required adjustments.

#### Job Title

In [11]:
df['Job Title'].value_counts()

Data Analyst                                 405
Senior Data Analyst                           90
Junior Data Analyst                           30
Business Data Analyst                         28
Sr. Data Analyst                              21
                                            ... 
Data Remediation Analyst                       1
Data Analyst, Center for Developing Brain      1
System Analyst (Data) - PL/SQl Server          1
Upstart - Data Analyst                         1
Celonis Data Engineer & Analyst                1
Name: Job Title, Length: 1272, dtype: int64

In [12]:
df['Job Title'].value_counts()[:20]

Data Analyst               405
Senior Data Analyst         90
Junior Data Analyst         30
Business Data Analyst       28
Sr. Data Analyst            21
Data Analyst Junior         17
Data Analyst II             17
Data Quality Analyst        17
Data Governance Analyst     16
Lead Data Analyst           15
Data Reporting Analyst      13
Financial Data Analyst      12
Data Analyst III            11
Data Analyst I              11
Marketing Data Analyst       9
Sr Data Analyst              9
Data Management Analyst      8
Data Warehouse Analyst       8
Technical Data Analyst       7
SQL Data Analyst             7
Name: Job Title, dtype: int64

- As we have seen some of the naming in the job title can be renamed (such as Sr. --> Senior, Jr. ---> Junior)

In [13]:
df['Job Title']= df['Job Title'].str.replace ('Sr. Data Analyst','Senior Data Analyst')

In [14]:
df['Job Title']= df['Job Title'].str.replace ('Sr Data Analyst','Senior Data Analyst')

In [15]:
df['Job Title']= df['Job Title'].str.replace ('Data Analyst Senior','Senior Data Analyst')

In [16]:
df['Job Title']= df['Job Title'].str.replace ('Jr. Data Analyst','Junior Data Analyst')

In [17]:
df['Job Title']= df['Job Title'].str.replace ('Jr Data Analyst','Junior Data Analyst')

In [18]:
df['Job Title']= df['Job Title'].str.replace ('Data Analyst Junior','Junior Data Analyst')

In [19]:
df['Job Title'].value_counts()[:20]

Data Analyst               405
Senior Data Analyst        121
Junior Data Analyst         50
Business Data Analyst       28
Data Analyst II             17
Data Quality Analyst        17
Data Governance Analyst     16
Lead Data Analyst           15
Data Reporting Analyst      13
Financial Data Analyst      12
Data Analyst I              11
Data Analyst III            11
Marketing Data Analyst       9
Data Management Analyst      8
Data Warehouse Analyst       8
SQL Data Analyst             7
Technical Data Analyst       7
Data Science Analyst         7
Data Security Analyst        6
Clinical Data Analyst        6
Name: Job Title, dtype: int64

In [20]:
df['Job Title'].isnull().sum()

0

- So far seems OK.
- Let's move on to the **Salary Estimate**

### Salary Estimate

In [21]:
df['Salary Estimate']

0        $37K-$66K (Glassdoor est.)
1        $37K-$66K (Glassdoor est.)
2        $37K-$66K (Glassdoor est.)
3        $37K-$66K (Glassdoor est.)
4        $37K-$66K (Glassdoor est.)
                   ...             
2248    $78K-$104K (Glassdoor est.)
2249    $78K-$104K (Glassdoor est.)
2250    $78K-$104K (Glassdoor est.)
2251    $78K-$104K (Glassdoor est.)
2252    $78K-$104K (Glassdoor est.)
Name: Salary Estimate, Length: 2253, dtype: object

In [22]:
df['Salary Estimate'].isnull().sum()

1

OK we have two issue to give attention.

- Missing value. In our case, only 1, but since we will use salary estimate in our analysis in detail, we don't want any row without salary estimate. 

- Salary estimate is not what we expected, it is not in the numeric form, we have **$** sign, **text** and numbers in the form of range between two number.

##### **Dealing with missing value in the Salary Estimate**

- In addition to domain knowledge and expertise, there are tons of different ways to deal with the missing values.
- in this EDA, we will use just one of them.

In [23]:
# First let's find the row with the missing value in the salary estimate column

In [24]:
df[df['Salary Estimate'].isnull()]

Unnamed: 0                                  Job Title Salary Estimate  \
2149        2149  Configuration/Data Management Analyst III             NaN   

                                        Job Description  Rating  \
2149  Job Description\nPosition Title: Configuration...     4.4   

         Company Name  Location  Headquarters                 Size  Founded  \
2149  Protingent\n4.4  Kent, WA  Bellevue, WA  51 to 200 employees   2001.0   

      Type of ownership                Industry             Sector  \
2149  Company - Private  Staffing & Outsourcing  Business Services   

                       Revenue Competitors Easy Apply  
2149  $25 to $50 million (USD)         NaN        NaN

In [25]:
# let's see this company has any other advertisement

In [26]:
df[df['Company Name']=='Protingent\n4.4']

Unnamed: 0                                          Job Title  \
2123        2123  Engines Configuration and Data Management Analyst   
2149        2149          Configuration/Data Management Analyst III   

                  Salary Estimate  \
2123  $55K-$101K (Glassdoor est.)   
2149                          NaN   

                                        Job Description  Rating  \
2123  Job Description\nPosition Title: Engines Confi...     4.4   
2149  Job Description\nPosition Title: Configuration...     4.4   

         Company Name  Location  Headquarters                 Size  Founded  \
2123  Protingent\n4.4  Kent, WA  Bellevue, WA  51 to 200 employees   2001.0   
2149  Protingent\n4.4  Kent, WA  Bellevue, WA  51 to 200 employees   2001.0   

      Type of ownership                Industry             Sector  \
2123  Company - Private  Staffing & Outsourcing  Business Services   
2149  Company - Private  Staffing & Outsourcing  Business Services   

                       Revenue Competitors Easy Apply  
2123  $25 to $50 million (USD)         NaN        NaN  
2149  $25 to $50 million (USD)         NaN        NaN

In [27]:
# Ok, we have another advertisement from the same company. Let's see the similarities in the job description.

In [28]:
df['Job Description'][2123]

'Job Description\nPosition Title: Engines Configuration and Data Management Analyst\n\nPosition Description: Protingent Staffing has an exciting contract opportunity with our client in Kent, WA.\n\nJob Qualifications:\nAA/AS or BA/BS in a related technical or liberal arts discipline\n3+ years of direct product configuration management experience (Product structure, parts, drawings, documents and requirements management)\n2+ years of experience working in a highly regulated environment (aerospace, nuclear, government, medical device, etc.)\nExperience working with software related to configuration management of complex product\nWorking knowledge of configuration management principles and practices\nFamiliar with product lifecycle management (PLM) and/or data management tools\nPossess high attention to detail and strong organizational/coordination skills\nProficient in Microsoft Office applications (Word, Excel, PowerPoint).\nExperience in aerospace industry\nProficient in technical writ

In [29]:
df['Job Description'][2149]

'Job Description\nPosition Title: Configuration/Data Management Analyst III\n\nPosition Description: Protingent has an opportunity for a Configuration Analyst in Kent, WA.\n\nJob Qualifications:\nAA/AS or BA/BS in a related technical or liberal arts discipline\n5-7 years of direct product configuration management experience (Product structure, parts, drawings and requirements management experience highly desired).\n3+ years of experience working in a highly regulated environment (security systems, medical device, aerospace, etc.).\nDeep experience with world-class toolsets for configuration management in complex products\nProven ability to communicate effectively both in writing and verbally with staff at all levels of the organization and external stakeholders, including regulatory agencies, customers and suppliers.\nExcellent technical writing and presentation skills.\nProficient with configuration management principles and practices\nProficient with product lifecycle management (PLM

- From the above job descriptions, even though there are similarities in the job description, job in the missing value row, needs more qualifications than the previous one
Based on the given info we can not assume same salary estimate for the missing salary estimate  value.
- It is better to drop missing value row.

In [30]:
df.drop(2149, inplace=True)

In [31]:
df['Salary Estimate'].isnull().sum()

0

#### To get Numeric values from Salary Estimate

- First let's remember, what format we have in the salary estimate variable

In [32]:
df['Salary Estimate'].sample(3)

1290     $60K-$124K (Glassdoor est.)
691     $113K-$132K (Glassdoor est.)
543       $37K-$70K (Glassdoor est.)
Name: Salary Estimate, dtype: object

- We have quite a lot of useful information, and we should make use of it.

Let's get the numbers out of it and by using the numbers, let's make: 
- maximum salary column
- minimum salary column
- average salary column

In [33]:
df['Salary_minimum']= df['Salary Estimate'].str.lstrip('$').str[:3].str.replace('K','').str.strip().astype('float')

In [34]:
df['Salary_maximum'] = df['Salary Estimate'].str[6:10].str.replace('K','').str.lstrip('$').str.strip().astype('float')

In [35]:
df['Salary_average'] = (df['Salary_maximum']+df['Salary_minimum'])/2

- Let's see everthing is in order.

In [36]:
df[['Salary Estimate','Salary_minimum','Salary_maximum','Salary_average']].sample(10)

Salary Estimate  Salary_minimum  Salary_maximum  \
2       $37K-$66K (Glassdoor est.)            37.0            66.0   
32      $46K-$87K (Glassdoor est.)            46.0            87.0   
190    $60K-$110K (Glassdoor est.)            60.0           110.0   
1754    $35K-$67K (Glassdoor est.)            35.0            67.0   
1310    $34K-$61K (Glassdoor est.)            34.0            61.0   
919     $47K-$76K (Glassdoor est.)            47.0            76.0   
699   $113K-$132K (Glassdoor est.)           113.0           132.0   
743     $60K-$66K (Glassdoor est.)            60.0            66.0   
861    $97K-$129K (Glassdoor est.)            97.0           129.0   
118     $51K-$87K (Glassdoor est.)            51.0            87.0   

      Salary_average  
2               51.5  
32              66.5  
190             85.0  
1754            51.0  
1310            47.5  
919             61.5  
699            122.5  
743             63.0  
861            113.0  
118             69.0

- Seems quite OK

#### **Job Description**

In [37]:
df['Job Description'][0]

"Are you eager to roll up your sleeves and harness data to drive policy change? Do you enjoy sifting through complex datasets to illuminate trends and insights? Do you see yourself working for a values-driven organization with a vision to tackle the most pressing injustices of our day?\n\nWe are looking to hire a bright, hard-working, and creative individual with strong data management skills and a demonstrated commitment to immigrant's rights. The Data Analyst will assist with analysis and reporting needs for Veras Center on Immigration and Justice (CIJ), working across its current projects and future Vera initiatives.\n\nWho we are:\n\nFounded in 1961, The Vera Institute is an independent, non-partisan, nonprofit organization that combines expertise in research, technical assistance, and demonstration projects to assist leaders in government and civil society examine justice policy and practice, and improve the systems people rely on for justice and safety.\nWe study problems that im

- Yes I agree quitea long one. But job descriptions contain tons of useful information. For this EDA we will focus on programming language requirements. Specifically, Python and SQL

In [38]:
df['Job Description'].isnull().sum()

0

- That's good, we can safely use job description variable for further analysis.

- First start with Python, and see how many job description contains/requires knowledge on Python.

In [39]:
df['python_job'] = df['Job Description'].str.contains('python', na=False, case=False)
df['python_job'].value_counts()

False    1615
True      637
Name: python_job, dtype: int64

- Then look at the SQL

In [40]:
df['SQL_job'] = df['Job Description'].str.contains('sql', na=False, case=False)
df['SQL_job'].value_counts()

True     1388
False     864
Name: SQL_job, dtype: int64

- Agreed good to see the old friend Ms Excel

In [41]:
df['excel_job'] = df['Job Description'].str.contains('excel', na=False, case=False)
df['excel_job'].value_counts()

True     1353
False     899
Name: excel_job, dtype: int64

- And lastly see Tableau

In [42]:
df['tableau_job'] = df['Job Description'].str.contains('tableau', na=False, case=False)
df['tableau_job'].value_counts()

False    1632
True      620
Name: tableau_job, dtype: int64

- Let's move on to the 'Rating'

#### **Ratings**

In [43]:
df['Rating'].sample(10)

1548    3.8
1825    NaN
2241    2.5
704     3.4
466     5.0
594     4.5
2003    2.8
739     4.7
2105    4.3
982     3.2
Name: Rating, dtype: float64

In [44]:
df['Rating'].isnull().sum()

272

- Right now, we will keep the row which has missing value on Rating, we ant to use salary and job title info on these rows.

#### **Company Name**

In [45]:
df['Company Name'].sample(10)

88      Global Atlantic Financial Group\n4.5
2242              The Judge Group, Inc.\n3.5
798                               Mondo\n3.9
2174                        Xcel Energy\n3.5
847                         Tempus Labs\n3.2
1243              Peyton Resource Group\n3.9
1285                    The RiverStone Group
319                    YAI/NIPD Network\n3.1
1267                               Stem\n3.2
915            National Computing Group\n4.7
Name: Company Name, dtype: object

- OK. Something weird with the company names, it includes number, which most probably reflects, their rating score.
- Let explore it.

In [46]:
df[['Company Name', 'Rating']].sample(5)

Company Name  Rating
2015                      OKCoin\n4.5     4.5
1034        Infotree Service Inc\n3.7     3.7
1104          Tandigm Health LLC\n4.2     4.2
1399  Lightbeam Health Solutions\n4.0     4.0
652                  Easyfind Company     NaN

- Yep, so we can safely remove rating part from the company name.

In [47]:
df['Company Name']= df['Company Name'].str.split('\n').str[0]

df['Company Name'].sample(3)

1633    Infotree Service Inc
916               HealthHelp
908                      FIS
Name: Company Name, dtype: object

- Let's see whether any missing value in company name

In [48]:
df['Company Name'].isnull().sum()

1

In [49]:
df[df['Company Name'].isnull()]

Unnamed: 0     Job Title             Salary Estimate  \
1860        1860  Data Analyst  $53K-$99K (Glassdoor est.)   

                                        Job Description  Rating Company Name  \
1860  Kindred at Home, part of the Kindred at Home f...     NaN          NaN   

             Location Headquarters Size  Founded  ... Revenue Competitors  \
1860  Mooresville, NC          NaN  NaN      NaN  ...     NaN         NaN   

     Easy Apply Salary_minimum Salary_maximum Salary_average  python_job  \
1860        NaN           53.0           99.0           76.0       False   

      SQL_job  excel_job  tableau_job  
1860     True       True        False  

[1 rows x 23 columns]

- We can keep this row, for job title, salary and job description

In [50]:
df['Industry'].value_counts()

IT Services                         325
Staffing & Outsourcing              322
Health Care Services & Hospitals    151
Consulting                          111
Computer Hardware & Software        111
                                   ... 
Trucking                              1
Express Delivery Services             1
Truck Rental & Leasing                1
Metals Brokers                        1
News Outlet                           1
Name: Industry, Length: 88, dtype: int64

- Seems OK.
- Let's see the missing values, if any.

In [51]:
df['Industry'].isnull().sum()

353

- We will keep these rows which has missing value on Industry.

#### Sector

In [52]:
df['Sector'].value_counts()

Information Technology                570
Business Services                     523
Finance                               169
Health Care                           151
Education                              52
Insurance                              51
Accounting & Legal                     43
Media                                  42
Manufacturing                          40
Retail                                 38
Government                             36
Biotech & Pharmaceuticals              33
Non-Profit                             26
Aerospace & Defense                    22
Transportation & Logistics             20
Construction, Repair & Maintenance     16
Consumer Services                      14
Oil, Gas, Energy & Utilities           13
Real Estate                            12
Telecommunications                     11
Restaurants, Bars & Food Services       8
Arts, Entertainment & Recreation        7
Travel & Tourism                        1
Mining & Metals                   

- Seems quite OK.
- Let's look at the missing values, at least we know one, but let's see any other missng value in this variable

In [53]:
df['Sector'].isnull().sum()

353

- We will keep these rows which has missing value on Sector.

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2252 entries, 0 to 2252
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         2252 non-null   int64  
 1   Job Title          2252 non-null   object 
 2   Salary Estimate    2252 non-null   object 
 3   Job Description    2252 non-null   object 
 4   Rating             1980 non-null   float64
 5   Company Name       2251 non-null   object 
 6   Location           2252 non-null   object 
 7   Headquarters       2080 non-null   object 
 8   Size               2089 non-null   object 
 9   Founded            1592 non-null   float64
 10  Type of ownership  2089 non-null   object 
 11  Industry           1899 non-null   object 
 12  Sector             1899 non-null   object 
 13  Revenue            2089 non-null   object 
 14  Competitors        521 non-null    object 
 15  Easy Apply         80 non-null     object 
 16  Salary_minimum     2252 

### Analysis Part

Let's start the analysis part with needed variables, namely:
- Job title
- Rating
- Company Name
- Industry
- Sector
- Salary Minimum
- Salary Maximum
- Salary Average
- Python job
- SQL job
- Excel job
- Tableau job

In [55]:
df.columns

Index(['Unnamed: 0', 'Job Title', 'Salary Estimate', 'Job Description',
       'Rating', 'Company Name', 'Location', 'Headquarters', 'Size', 'Founded',
       'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors',
       'Easy Apply', 'Salary_minimum', 'Salary_maximum', 'Salary_average',
       'python_job', 'SQL_job', 'excel_job', 'tableau_job'],
      dtype='object')

In [56]:
df_analyst = df[['Job Title', 'Company Name', 'Rating', 'Industry', 'Sector', 'Salary_minimum','Salary_maximum', 'Salary_average','python_job', 'SQL_job', 'excel_job', 'tableau_job']]
df_analyst.head()

Job Title  \
0  Data Analyst, Center on Immigration and Justic...   
1                               Quality Data Analyst   
2  Senior Data Analyst, Insights & Analytics Team...   
3                                       Data Analyst   
4                             Reporting Data Analyst   

                         Company Name  Rating  \
0           Vera Institute of Justice     3.2   
1  Visiting Nurse Service of New York     3.8   
2                         Squarespace     3.4   
3                            Celerity     4.1   
4                             FanDuel     3.9   

                           Industry                            Sector  \
0                 Social Assistance                        Non-Profit   
1  Health Care Services & Hospitals                       Health Care   
2                          Internet            Information Technology   
3                       IT Services            Information Technology   
4               Sports & Recreation  Arts, Entertainment & Recreation   

   Salary_minimum  Salary_maximum  Salary_average  python_job  SQL_job  \
0            37.0            66.0            51.5        True     True   
1            37.0            66.0            51.5       False     True   
2            37.0            66.0            51.5        True     True   
3            37.0            66.0            51.5       False     True   
4            37.0            66.0            51.5        True     True   

   excel_job  tableau_job  
0      False        False  
1       True        False  
2       True         True  
3      False         True  
4       True        False

In [57]:
df_analyst.isnull().sum()

Job Title           0
Company Name        1
Rating            272
Industry          353
Sector            353
Salary_minimum      0
Salary_maximum      0
Salary_average      0
python_job          0
SQL_job             0
excel_job           0
tableau_job         0
dtype: int64

In [58]:
df_analyst.describe()

Rating  Salary_minimum  Salary_maximum  Salary_average
count  1980.000000     2252.000000     2252.000000     2252.000000
mean      3.731566       54.266874       89.979130       72.123002
std       0.670333       19.575375       29.315605       23.600734
min       1.000000       24.000000       38.000000       33.500000
25%       3.300000       41.000000       70.000000       58.000000
50%       3.700000       50.000000       87.000000       69.000000
75%       4.100000       64.000000      104.000000       80.500000
max       5.000000      113.000000      190.000000      150.000000

Let's look at the some of the information we can get from above table:

- Average and also median value for the ratings of the companies very close to 4 (at round 3.7)
- Average minimum salary is around 54K and median value for minimum salary is around 50K. We can expect outliers from salary distribution and we acn also expect right skewed distribution of the minimum salary.
- Average maximum salary is almost 90K (89.97) and median value for maximum salary is 87K. We can expect outliers from salary distribution and we can also expect right skewed distribution of the maximum salary.
- Average salary is around 72K, but still we can expect several outliers for this variable(min= 33.5K, max= 150K)

In [59]:
fig = px.histogram(df_analyst, x= 'Salary_minimum', title='Minimum Salary of Data Analyst Jobs', marginal="box",hover_data = df_analyst[['Job Title', 'Company Name']])

fig.show()

- As seen in the histogram, minimum salary for data analyst is an average between 40-70K, but we have quite skewed distribution on the minimum salary.

In [60]:
fig = px.histogram(df_analyst, x= 'Salary_maximum', title='Maximum Salary of Data Analyst Jobs', marginal="box",hover_data = df_analyst[['Job Title', 'Company Name']])

fig.show()

- As seen in the histogram, maximum salary for data analyst is an average between 65-90K, but we have quite skewed distribution on the maximum salary.

In [61]:
fig = px.histogram(df_analyst, x= 'Salary_average', title='Average Salary of Data Analyst Jobs',marginal="box",hover_data = df_analyst[['Job Title', 'Company Name']])

fig.show()

- As seen in the histogram, average salary for data analyst is an average between 60-80K, but we have skewed distribution on the average salary.

- OK, we have overall info about the salary of the all data analyst job advertisements.
- Next we will specifically see 'Data Analyst' job advertisement and it's salary scale

### Data Analyst

In [62]:
data_analyst_df = df_analyst[df_analyst['Job Title']=='Data Analyst']
data_analyst_df.head()

Job Title        Company Name  Rating  \
3   Data Analyst            Celerity     4.1   
5   Data Analyst             Point72     3.9   
8   Data Analyst                DMGT     4.0   
10  Data Analyst  NYU Langone Health     4.0   
11  Data Analyst            BulbHead     NaN   

                                 Industry                  Sector  \
3                             IT Services  Information Technology   
5   Investment Banking & Asset Management                 Finance   
8        Venture Capital & Private Equity                 Finance   
10       Health Care Services & Hospitals             Health Care   
11                                    NaN                     NaN   

    Salary_minimum  Salary_maximum  Salary_average  python_job  SQL_job  \
3             37.0            66.0            51.5       False     True   
5             37.0            66.0            51.5        True     True   
8             37.0            66.0            51.5        True     True   
10            37.0            66.0            51.5       False    False   
11            37.0            66.0            51.5        True     True   

    excel_job  tableau_job  
3       False         True  
5        True        False  
8        True        False  
10       True        False  
11       True         True

In [63]:
data_analyst_df.describe()

Rating  Salary_minimum  Salary_maximum  Salary_average
count  349.000000      405.000000      405.000000      405.000000
mean     3.850143       54.296296       90.874074       72.585185
std      0.679529       21.089181       35.892185       28.007461
min      1.000000       24.000000       38.000000       33.500000
25%      3.400000       38.000000       67.000000       53.500000
50%      3.900000       51.000000       85.000000       68.000000
75%      4.300000       60.000000      103.000000       79.500000
max      5.000000      110.000000      190.000000      150.000000

Let's look at the some of the information we can get from above table:

- Average and also median value for the ratings of the companies very close to 4 (at round 3.9), it is a little bit more than overall job advertisements' rating average (3,7)
- Average minimum salary is around 54K and median value for minimum salary is 51K. It is almost same with the whole data minimum salary mean and median. We can expect outliers from salary distribution and we acn also expect right skewed distribution of the minimum salary.
- Average maximum salary is almost 90K (89.97) and median value for maximum salary is 85K. MMedian value is less than whole data maximum salary median value. We can expect outliers from salary distribution and we can also expect right skewed distribution of the maximum salary.
- Average salary is around 72K, but still we can expect several outliers for this variable(min= 33.5K, max= 150K)

In [64]:

fig = px.histogram(data_analyst_df, x= 'Salary_minimum',title='Minimum Salary of Data Analyst', marginal="box",hover_data = data_analyst_df[['Job Title', 'python_job', 'SQL_job','excel_job','tableau_job']])
fig.show()

- As seen in the graph, minimum salary for data analyst is an average between 40-60K, but we have quite skewed distribution on the minimum salary.

In [65]:
fig = px.histogram(data_analyst_df, x= 'Salary_maximum', title='Maximum Salary of Data Analyst',marginal="box",hover_data = data_analyst_df[['Job Title', 'python_job', 'SQL_job','excel_job','tableau_job']])
fig.show()

- As seen in the histogram, maximum salary for data analyst is an average between 67-95K, but we have quite skewed distribution on the maximum salary.

In [66]:
fig = px.histogram(data_analyst_df, x= 'Salary_average',title='Average Salary of Data Analyst', marginal="box",hover_data = data_analyst_df[['Job Title', 'python_job', 'SQL_job','excel_job','tableau_job']])
fig.show()

- As seen in the histogram, average salary for data analyst is an average between 55-80K, but we have skewed distribution on the average salary.

### Job Openings by Job Title

- Let's see top 20 job title in the data analyst job advertisements.

In [67]:
df_analyst['Job Title'].value_counts()[:20]

Data Analyst               405
Senior Data Analyst        121
Junior Data Analyst         50
Business Data Analyst       28
Data Analyst II             17
Data Quality Analyst        17
Data Governance Analyst     16
Lead Data Analyst           15
Data Reporting Analyst      13
Financial Data Analyst      12
Data Analyst I              11
Data Analyst III            11
Marketing Data Analyst       9
Data Management Analyst      8
Data Warehouse Analyst       8
Data Science Analyst         7
Technical Data Analyst       7
SQL Data Analyst             7
Data Security Analyst        6
Research Data Analyst        6
Name: Job Title, dtype: int64

In [68]:
fig=go.Figure()
fig.add_trace(go.Scatter(
    x= df_analyst['Job Title'].value_counts()[:20].index,
    y= df_analyst['Job Title'].value_counts()[:20].values,
    name='Number of Job Openings',
    mode='markers+text+lines',
    marker_color='blue',
    marker_size=10,
    text=df_analyst['Job Title'].value_counts()[:20].values,
    textposition='top center',
    line=dict(color='red',dash='dash'),
))
fig.update_layout(
    title= "<b>Number of Job Openings by Job titles</b>",
    xaxis_title="<b>Job Titles</b>",
    yaxis_title="<b>Number of Job Openings</b>",
    template='seaborn',
    font=dict(
        size=12,
        color="Black",
        family="Oswald', sans-serif"
        ),
    xaxis=dict(showgrid=True),
    yaxis=dict(showgrid=True),
    yaxis2=dict(showgrid=True,overlaying='y',side='right',title='<b>Number of Job Openings</b>'),
    legend=dict(yanchor="top",
    y=1.3,
    xanchor="left",
    x=0.78)
)
fig.show()

As shown in the plot;
- Data Analyst
- Senior Data Analyst
- Junior Data Analyst
- Business Data Analyst are the most used titles in the job advertisements.

### Job Openings by Industry

In [69]:
df_analyst['Industry'].value_counts()[:20]

IT Services                                325
Staffing & Outsourcing                     322
Health Care Services & Hospitals           151
Consulting                                 111
Computer Hardware & Software               111
Investment Banking & Asset Management       78
Enterprise Software & Network Solutions     69
Internet                                    65
Banks & Credit Unions                       51
Advertising & Marketing                     51
Insurance Carriers                          47
Colleges & Universities                     40
Accounting                                  39
Biotech & Pharmaceuticals                   33
Social Assistance                           23
Aerospace & Defense                         22
Video Games                                 20
State & Regional Agencies                   15
Lending                                     15
Logistics & Supply Chain                    13
Name: Industry, dtype: int64

In [70]:
fig=go.Figure()
fig.add_trace(go.Scatter(
    x= df_analyst['Industry'].value_counts()[:20].index,
    y= df_analyst['Industry'].value_counts()[:20].values,
    name='Number of Job Openings',
    mode='markers+text+lines',
    marker_color='red',
    marker_size=10,
    text=df_analyst['Industry'].value_counts()[:20].values,
    textposition='top center',
    line=dict(color='white',dash='dash'),
))
fig.update_layout(
    title= "<b>Number of Job Openings by Industry</b>",
    xaxis_title="<b>industry</b>",
    yaxis_title="<b>Number of Job Openings</b>",
    template='seaborn',
    font=dict(
        size=12,
        color="Black",
        family="Oswald', sans-serif"
        ),
    xaxis=dict(showgrid=False),
    yaxis=dict(showgrid=False),
    yaxis2=dict(showgrid=True,overlaying='y',side='right',title='<b>Number of Job Openings</b>'),
    legend=dict(yanchor="top",
    y=1.3,
    xanchor="left",
    x=0.78)
)
fig.show()

As shown in the plot;

- IT Services                               
- Staffing & Outsourcing                     
- Health Care Services & Hospitals           
- Consulting                                 
- Computer Hardware & Software   are the most data analyst job opening advertised industries.

- It is important to mention also, Banking (Investment Banks - Banks) also advertised 129 (combined)  data analyst job  openings

### Job Openings by Sector

In [71]:
df_analyst['Sector'].value_counts()[:20]

Information Technology                570
Business Services                     523
Finance                               169
Health Care                           151
Education                              52
Insurance                              51
Accounting & Legal                     43
Media                                  42
Manufacturing                          40
Retail                                 38
Government                             36
Biotech & Pharmaceuticals              33
Non-Profit                             26
Aerospace & Defense                    22
Transportation & Logistics             20
Construction, Repair & Maintenance     16
Consumer Services                      14
Oil, Gas, Energy & Utilities           13
Real Estate                            12
Telecommunications                     11
Name: Sector, dtype: int64

In [72]:
fig=go.Figure()
fig.add_trace(go.Scatter(
    x= df_analyst['Sector'].value_counts()[:20].index,
    y= df_analyst['Sector'].value_counts()[:20].values,
    name='Number of Job Openings',
    mode='markers+text+lines',
    marker_color='blue',
    marker_size=10,
    text=df_analyst['Sector'].value_counts()[:20].values,
    textposition='top center',
    line=dict(color='white',dash='dash'),
))
fig.update_layout(
    title= "<b>Number of Job Openings by Sector</b>",
    xaxis_title="<b>Sector</b>",
    yaxis_title="<b>Number of Job Openings</b>",
    template='seaborn',
    font=dict(
        size=12,
        color="Black",
        family="Oswald', sans-serif"
        ),
    xaxis=dict(showgrid=True),
    yaxis=dict(showgrid=True),
    yaxis2=dict(showgrid=True,overlaying='y',side='right',title='<b>Number of Job Openings</b>'),
    legend=dict(yanchor="top",
    y=1.3,
    xanchor="left",
    x=0.78)
)
fig.show()

As shown in the plot;

- Information Technology                               
- Business Services
- Finance
- Health Care are the most data analyst job opening advertised sectors.


- Now let's see **Python, SQL, Excel, Tableau** in an action.

### Required Programming Languages & Program Skills by Job Title

In [73]:
df4 = df_analyst[['Job Title','python_job', 'SQL_job','excel_job','tableau_job']].copy()

Lang = df4.groupby('Job Title')[['python_job', 'SQL_job','excel_job','tableau_job']].sum().sort_values(by='python_job',ascending=False).head(10)
df_lang = pd.DataFrame(Lang)
df_lang = df_lang.reset_index()

df_lang['number_of_job_openings'] = df_analyst['Job Title'].value_counts()[:10].values
columnsTitles = ['Job Title', 'number_of_job_openings','python_job', 'SQL_job','excel_job','tableau_job']

df_lang = df_lang.reindex(columns=columnsTitles)

df_lang


Job Title  number_of_job_openings  \
0                                       Data Analyst                     405   
1                                Senior Data Analyst                     121   
2                              Business Data Analyst                      50   
3               NY Healthcare Data/Reporting Analyst                      28   
4                             Marketing Data Analyst                      17   
5            Data Analyst III (Healthcare Analytics)                      17   
6                                  Lead Data Analyst                      16   
7               TX Healthcare Data/Reporting Analyst                      15   
8  Data Analyst, Data & Analytics (Advanced Analy...                      13   
9                  Healthcare Data/Reporting Analyst                      12   

   python_job  SQL_job  excel_job  tableau_job  
0         125      301        239          110  
1          68      104         67           64  
2          11       21         17           11  
3           5        5          5            5  
4           5        9          6            5  
5           5        5          5            5  
6           5       13          7           10  
7           4        4          4            4  
8           4        4          0            4  
9           4        4          4            4

- Based on the job advertisements' requirements, we can safely assume that, SQL and Excel keeps their importance. As a programming language Python is required almost 1 out of 3 times in the job advertisemenst, same is also true for visualiztion tool Tableau.

In [74]:
fig = px.bar(df_lang, x='Job Title', y=['python_job', 'SQL_job','excel_job','tableau_job'], title="Languages")
fig.show()

- Finally see  based on company name, different salary distributions.

### Data Analyst Jobs Salary Range based on  Company Name with Rating Scores

In [75]:
fig = px.scatter(df_analyst, x="Salary_minimum", y="Company Name", 
                 color="Rating", 
                 hover_data=['Industry', 'Job Title'], 
                 title = "Minimum Salary by Company Name with Rating Scores")
fig.show()

- Minimum salary 110K is quite, I mean quite OK for working 3.9 rating company, like Netflix :)

In [76]:
fig = px.scatter(df_analyst, x="Salary_maximum", y="Company Name", 
                 color="Rating", 
                 hover_data=['Industry', 'Job Title'], 
                 title = "Maximum Salary by Company Name with Rating Scores")
fig.show()

- Company name 'Enjoy', maximum salary offfer 190K, most probably I will also 'Enjoy' to work there.

In [77]:
fig = px.scatter(df_analyst, x="Salary_average", y="Company Name", 
                 color="Rating", 
                 hover_data=['Industry', 'Job Title'], 
                 title = "Average Salary by Company Name with Rating Scores")
fig.show()

- It was a quite pleasure to share with you this detailed, beginner friendly EDA. Thanks for your time. 

- All the best 